File: Letter Combinations of a Phone Number - Leetcode 17.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits == "":
            return []

        ans, sol = [], []
        
        letter_map = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }
        
        n = len(digits)

        def backtrack(i=0):
            if i == n:
                ans.append("".join(sol))
                return

            for letter in letter_map[digits[i]]:
                sol.append(letter)
                backtrack(i + 1)
                sol.pop()

        backtrack(0)
        return ans

# Time Complexity: O(n * 4^n)
# Space Complexity: O(n)

File: Letter Combinations of a Phone Number - Leetcode 17/Letter Combinations of a Phone Number - Leetcode 17.cpp
#include <vector>
#include <string>
#include <unordered_map>

class Solution {
public:
    std::vector<std::string> letterCombinations(std::string digits) {
        std::vector<std::string> ans;
        if (digits.empty()) return ans;

        std::unordered_map<char, std::string> letterMap = {
            {'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"},
            {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}
        };
        
        std::string path;
        backtrack(digits, 0, path, letterMap, ans);
        return ans;
    }

private:
    void backtrack(const std::string& digits, int index, std::string& path,
                   const std::unordered_map<char, std::string>& letterMap,
                   std::vector<std::string>& ans) {
        if (index == digits.size()) {
            ans.push_back(path);
            return;
        }

        const std::string& letters = letterMap.at(digits[index]);
        for (char letter : letters) {
            path.push_back(letter);
            backtrack(digits, index + 1, path, letterMap, ans);
            path.pop_back();
        }
    }
};
